spec/unit/actions/collect_data_product_spec.rb

```diff
@@ -55,4 +55,18 @@
       expect(result[:params][:data_product]).to be_a GoodData::DataProduct
     end
   end
+
+  context 'when the passed data_product does not exist' do
+    let(:data_product_id) { 'non-existing-data-product' }
+    include_context 'data_product config'
+
+    before do
+      allow(domain).to receive(:data_products) { fail RestClient::BadRequest }
+    end
+
+    it 'creates the data product' do
+      expect(domain).to receive(:create_data_product)
```

Comment:
It would be nice to also check parameters of create_data_product. Something like receive(:create_data_product).with(id: expected_id)